openexr-rs/src/core/preview_image.rs

```diff
@@ -1,23 +1,470 @@
+            std::slice::from_raw_parts_mut(
+                pixels as *mut PreviewRgba,
+                pixel_count,
+            )
+        }
+    }
+
+    /// Sets the pixel value at the specified coordinates
+    ///
+    /// ## Errors
+    /// * [`Error::OutOfRange`] - If the coordinates are outside of the image size
+    /// * [`Error::Base`] - If an error occurs
+    ///
+    pub fn set_pixel(&mut self, x: u32, y: u32, value: PreviewRgba) -> Result<()> {
+        unsafe {
+            let mut pixel = std::ptr::null_mut();
+
+            if x >= self.width() || y >= self.height() {
+                return Err(Error::OutOfRange);
+            }
+
+            sys::Imf_PreviewImage_pixel(self.0, &mut pixel, x, y)
+                .into_result()?;
+
+            *(pixel as *mut PreviewRgba) = value
+        }
+
+        Ok(())
+    }
+
+    /// Gets the pixel value at the specified coordinates
+    ///
+    /// Returns `None` if the coordinates are outside of the image size,
+    /// or some other error ocurred
+    ///
+    pub fn get_pixel(&self, x: u32, y: u32) -> Option<PreviewRgba> {
+        unsafe {
+            let pixel = std::ptr::null_mut();
+
+            if x >= self.width() || y >= self.height() {
+                return None;
+            }
+
+            match sys::Imf_PreviewImage_pixel_const(self.0, pixel, x, y)
+                .into_result()
+            {
+                Ok(_) => Some(*(pixel as *const PreviewRgba)),
+                Err(_) => None,
+            }
+        }
+    }
+}
+
+impl Clone for PreviewImage {
+    fn clone(&self) -> Self {
+        let mut _inner = std::ptr::null_mut();
+        unsafe {
+            sys::Imf_PreviewImage_copy(&mut _inner, self.0)
+                .into_result()
+                .expect("Error invoking copy ctor");
+        }
+
+        PreviewImage(_inner)
+    }
+}
+
+impl Drop for PreviewImage {
+    fn drop(&mut self) {
+        unsafe {
+            sys::Imf_PreviewImage_dtor(self.0);
+        }
+    }
+}
+
+/// Holds the value of a PreviewImage pixel
+///
+/// Intensity is proportional to pow(x/255, 2.2) for r, g, and b components.
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct PreviewRgba {
+    /// Red component of the pixel's color.
+    /// Intensity is proportional to pow(x/255, 2.2).
+    r: u8,
+
+    /// Green component of the pixel's color.
+    /// Intensity is proportional to pow(x/255, 2.2).
+    g: u8,
+
+    /// Blue component of the pixel's color.
+    /// Intensity is proportional to pow(x/255, 2.2).
+    b: u8,
+
+    /// The pixel's alpha.
+    /// 0 == transparent, 255 == opaque.
+    a: u8,
+}
+
+impl PreviewRgba {
+    /// Creates a `PreviewRgba` from RGB and Alpha `u8` values
+    ///
+    pub fn from_u8(r: u8, g: u8, b: u8, a: u8) -> PreviewRgba {
+        PreviewRgba { r, g, b, a }
+    }
+    /// Creates a `PreviewRgba` from RGB and Alpha `u8` values
+    ///
+    pub fn from_f16(r: f16, g: f16, b: f16, a: f16) -> PreviewRgba {

```

Comment:
We should apply gamma correction here as f16 values will be assumed to be linear. Also, would this work better as a From<Rgba> impl?

Reply:
What do you have in mind for the `From`? `impl From<Rgba> for PreviewRgba` or something else? And should we still keep this function?

Reply:
Added gamma correction using the formula extracted from OpenEXR C++ source code and also added `from_f16_linear` and `from_f32_linear` that convert directly without applying any gamma correction.